# Polars

## Задание 1: Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью polars - 1 балл

In [1]:
import polars as pl
df_train = pl.read_csv("train.csv")

## Задание 2: Выведите основную информацию о датасете: информацию о типах данных, число пропусков, средние значения и т.д. - 1 балл

In [2]:
df_train.describe()

shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ statistic  ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---      ┆   ┆ ---       ┆ ---       ┆ ---   ┆ ---      │
│ str        ┆ f64         ┆ f64      ┆ f64      ┆   ┆ str       ┆ f64       ┆ str   ┆ str      │
╞════════════╪═════════════╪══════════╪══════════╪═══╪═══════════╪═══════════╪═══════╪══════════╡
│ count      ┆ 891.0       ┆ 891.0    ┆ 891.0    ┆ … ┆ 891       ┆ 891.0     ┆ 204   ┆ 889      │
│ null_count ┆ 0.0         ┆ 0.0      ┆ 0.0      ┆ … ┆ 0         ┆ 0.0       ┆ 687   ┆ 2        │
│ mean       ┆ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ … ┆ null      ┆ 32.204208 ┆ null  ┆ null     │
│ std        ┆ 257.353842  ┆ 0.486592 ┆ 0.836071 ┆ … ┆ null      ┆ 49.693429 ┆ null  ┆ null     │
│ min        ┆ 1.0         ┆ 0.0      ┆ 1.0      ┆ … ┆ 110152    ┆ 0.0       ┆ A10   ┆ C        │
│ 25%        ┆ 224.0       ┆ 0.0      ┆ 2.0      ┆ … ┆ null      ┆ 7.925     ┆ null  ┆ null     │
│ 50%        ┆ 446.0       ┆ 0.0      ┆ 3.0      ┆ … ┆ null      ┆ 14.4542   ┆ null  ┆ null     │
│ 75%        ┆ 669.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ null      ┆ 31.0      ┆ null  ┆ null     │
│ max        ┆ 891.0       ┆ 1.0      ┆ 3.0      ┆ … ┆ WE/P 5735 ┆ 512.3292  ┆ T     ┆ S        │
└────────────┴─────────────┴──────────┴──────────┴───┴───────────┴───────────┴───────┴──────────┘

In [3]:
df_train.dtypes

[Int64,
 Int64,
 Int64,
 String,
 String,
 Float64,
 Int64,
 Int64,
 String,
 Float64,
 String,
 String]

In [4]:
df_train.null_count()

shape: (1, 12)
┌────────────┬──────────┬────────┬──────┬───┬────────┬──────┬───────┬──────────┐
│ PassengerI ┆ Survived ┆ Pclass ┆ Name ┆ … ┆ Ticket ┆ Fare ┆ Cabin ┆ Embarked │
│ d          ┆ ---      ┆ ---    ┆ ---  ┆   ┆ ---    ┆ ---  ┆ ---   ┆ ---      │
│ ---        ┆ u32      ┆ u32    ┆ u32  ┆   ┆ u32    ┆ u32  ┆ u32   ┆ u32      │
│ u32        ┆          ┆        ┆      ┆   ┆        ┆      ┆       ┆          │
╞════════════╪══════════╪════════╪══════╪═══╪════════╪══════╪═══════╪══════════╡
│ 0          ┆ 0        ┆ 0      ┆ 0    ┆ … ┆ 0      ┆ 0    ┆ 687   ┆ 2        │
└────────────┴──────────┴────────┴──────┴───┴────────┴──────┴───────┴──────────┘

In [43]:
df_train.mean()

shape: (1, 12)
┌─────────────┬──────────┬──────────┬──────┬───┬────────┬───────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass   ┆ Name ┆ … ┆ Ticket ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---      ┆ ---  ┆   ┆ ---    ┆ ---       ┆ ---   ┆ ---      │
│ f64         ┆ f64      ┆ f64      ┆ str  ┆   ┆ str    ┆ f64       ┆ str   ┆ str      │
╞═════════════╪══════════╪══════════╪══════╪═══╪════════╪═══════════╪═══════╪══════════╡
│ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ null ┆ … ┆ null   ┆ 32.204208 ┆ null  ┆ null     │
└─────────────┴──────────┴──────────┴──────┴───┴────────┴───────────┴───────┴──────────┘

## Задание 3: Посчитайте количество пассажиров каждого класса (Pclass) - 1 балл

In [3]:
pclass_counts = df_train.get_column("Pclass")

In [4]:
pclass_counts.value_counts()

shape: (3, 2)
┌────────┬───────┐
│ Pclass ┆ count │
│ ---    ┆ ---   │
│ i64    ┆ u32   │
╞════════╪═══════╡
│ 3      ┆ 491   │
│ 2      ┆ 184   │
│ 1      ┆ 216   │
└────────┴───────┘

## Задание 4: Выведите количество выживших мужчин и женщин на корабле - 1 балл

In [46]:
df_train.group_by('Sex').agg(pl.col('Survived').sum())

shape: (2, 2)
┌────────┬──────────┐
│ Sex    ┆ Survived │
│ ---    ┆ ---      │
│ str    ┆ i64      │
╞════════╪══════════╡
│ female ┆ 233      │
│ male   ┆ 109      │
└────────┴──────────┘

## Задание 5: Выведите часть таблицы с пассажирами, возраст которых больше 44 лет - 1 балл

In [8]:
df_train.filter(pl.col('Age') > 44)

shape: (115, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16.0    ┆ null  ┆ S        │
│             ┆          ┆        ┆ (Mary D          ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ Kingcome…        ┆   ┆            ┆         ┆       ┆          │
│ 34          ┆ 0        ┆ 2      ┆ Wheadon, Mr.     ┆ … ┆ C.A. 24579 ┆ 10.5    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Edward H         ┆   ┆            ┆         ┆       ┆          │
│ 53          ┆ 1        ┆ 1      ┆ Harper, Mrs.     ┆ … ┆ PC 17572   ┆ 76.7292 ┆ D33   ┆ C        │
│             ┆          ┆        ┆ Henry Sleeper    ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (My…             ┆   ┆            ┆         ┆       ┆          │
│ …           ┆ …        ┆ …      ┆ …                ┆ … ┆ …          ┆ …       ┆ …     ┆ …        │
│ 858         ┆ 1        ┆ 1      ┆ Daly, Mr. Peter  ┆ … ┆ 113055     ┆ 26.55   ┆ E17   ┆ S        │
│             ┆          ┆        ┆ Denis            ┆   ┆            ┆         ┆       ┆          │
│ 863         ┆ 1        ┆ 1      ┆ Swift, Mrs.      ┆ … ┆ 17466      ┆ 25.9292 ┆ D17   ┆ S        │
│             ┆          ┆        ┆ Frederick Joel   ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ (Ma…             ┆   ┆            ┆         ┆       ┆          │
│ 872         ┆ 1        ┆ 1      ┆ Beckwith, Mrs.   ┆ … ┆ 11751      ┆ 52.5542 ┆ D35   ┆ S        │
│             ┆          ┆        ┆ Richard Leonard… ┆   ┆            ┆         ┆       ┆          │
│ 874         ┆ 0        ┆ 3      ┆ Vander Cruyssen, ┆ … ┆ 345765     ┆ 9.0     ┆ null  ┆ S        │
│             ┆          ┆        ┆ Mr. Victor       ┆   ┆            ┆         ┆       ┆          │
│ 880         ┆ 1        ┆ 1      ┆ Potter, Mrs.     ┆ … ┆ 11767      ┆ 83.1583 ┆ C50   ┆ C        │
│             ┆          ┆        ┆ Thomas Jr (Lily  ┆   ┆            ┆         ┆       ┆          │
│             ┆          ┆        ┆ A…               ┆   ┆            ┆         ┆       ┆          │
└─────────────┴──────────┴────────┴──────────────────┴───┴────────────┴─────────┴───────┴──────────┘

# Ускорение работы с pandas

## Задание 1: Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью pandas

In [11]:
import pandas as pd
df_train_pd = pd.read_csv("train.csv")

## Задание 2: Посчитайте средний возраст пассажиров и его стандартное отклонение с помощью bottleneck - 1 бал

In [17]:
import bottleneck as bn
mean_age = bn.nanmean(df_train_pd.Age)
print(f"Средний возраст пассажиров: {mean_age}")

Средний возраст пассажиров: 29.69911764705882


In [18]:
std_age = bn.nanstd(df_train_pd.Age)
print(f"Средний возраст пассажиров: {std_age}")

Средний возраст пассажиров: 14.516321150817317


## Задание 3: Для каждого пассажира умножьте значение столбца Fare на 1.3 и сохраните результаты как новый столбец Fare_new - 1 балл

In [22]:
df_train_pd['Fare_new'] = [row.Fare * 1.3 for row in df_train_pd.itertuples()]
df_train_pd.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  Fare_new  
0      0         A/5 21171   7.2500   NaN        S   9.42500  
1      0          PC 17599  71.2833   C85        C  92.66829  
2      0  STON/O2. 3101282   7.9250   NaN        S  10.30250  
3      0            113803  53.1000  C123        S  69.03000  
4      0            373450   8.0500   NaN        S  10.46500  
5      0            330877   8.4583   NaN        Q  10.99579  
6      0             17463  51.8625   E46        S  67.42125  
7      1            349909  21.0750   NaN        S  27.39750  
8      2            347742  11.1333   NaN        S  14.47329  
9      0            237736  30.0708   NaN        C  39.09204

# Оптимизация типов pandas

# Задание 1: 1. Считайте датасет из файла Housing.csv (это данные о ценах домов) с помощью pandas

In [23]:
df_housing = pd.read_csv("Housing.csv")

## Задание 2: Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных - напишите свои выводы в комментариях - 2 балла

In [25]:
df_housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB


In [26]:
df_housing.describe()

price          area    bedrooms   bathrooms     stories  \
count  5.450000e+02    545.000000  545.000000  545.000000  545.000000   
mean   4.766729e+06   5150.541284    2.965138    1.286239    1.805505   
std    1.870440e+06   2170.141023    0.738064    0.502470    0.867492   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.430000e+06   3600.000000    2.000000    1.000000    1.000000   
50%    4.340000e+06   4600.000000    3.000000    1.000000    2.000000   
75%    5.740000e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

          parking  
count  545.000000  
mean     0.693578  
std      0.861586  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000

In [29]:
mainroad_values = df_housing['mainroad'].unique()
print(mainroad_values)

['yes' 'no']


In [30]:
guestroom_values = df_housing['guestroom'].unique()
print(guestroom_values)

['no' 'yes']


In [31]:
basement_values = df_housing['basement'].unique()
print(basement_values)

['no' 'yes']


In [32]:
hotwaterheating_values = df_housing['hotwaterheating'].unique()
print(hotwaterheating_values)

['no' 'yes']


In [33]:
airconditioning_values = df_housing['airconditioning'].unique()
print(airconditioning_values)

['yes' 'no']


In [34]:
prefarea_values = df_housing['prefarea'].unique()
print(prefarea_values)

['yes' 'no']


In [36]:
furnishingstatus_values = df_housing['furnishingstatus'].unique()
print(furnishingstatus_values)

['furnished' 'semi-furnished' 'unfurnished']


Диапазоны:
* **int8** (-128 до 127)
* **int16** (-32,768 до 32,767)
* **int32** (-2,147,483,648 до 2,147,483,647.)
* **int64** (-9,223,372,036,854,775,808 до 9,223,372,036,854,775,807)

Оптимизация:
* **price:** принимает значения от 1,750,000 до 13,300,000. Эти значения попадают в пределы диапазона **int32**, поэтому можем изменить тип с int64 на int32
* **area:** принимает значения от 1,650 до 16,200. Эти значения попадают в пределы диапазона **int16**, поэтому можем изменить тип с int64 на int16
* **bedrooms:** принимает значения от 1 до 6. Эти значения попадают в пределы диапазона **int8**, поэтому можем изменить тип с int64 на int8
* **bathrooms:** принимает значения от 1 до 4. Эти значения попадают в пределы диапазона **int8**, поэтому можем изменить тип с int64 на int8
* **stories:** принимает значения от 1 до 4. Эти значения попадают в пределы диапазона **int8**, поэтому можем изменить тип с int64 на int8
* **parking:** принимает значения от 0 до 3. Эти значения попадают в пределы диапазона **int8**, поэтому можем изменить тип с int64 на int8
* **mainroad, guestroom, basement, hotwaterheating, airconditioning, prefarea, furnishingstatus:** так как в этих переменных содержатся повторяющиеся однотипные элементы (yes/no; furnished/semi-furnished/unfurnished), то мы можем использовать переход к категориальным значениям 

## Задание 3: Поменяйте типы данных столбцов датафрейма на выбранные вами в прошлом пункте и сравните потребление памяти до и после оптимизации - 1 балл

In [41]:
import numpy as np
df_housing['price'] = df_housing['price'].astype(np.int32)
df_housing['area'] = df_housing['area'].astype(np.int16)
df_housing['bedrooms'] = df_housing['bedrooms'].astype(np.int8)
df_housing['bathrooms'] = df_housing['bathrooms'].astype(np.int8)
df_housing['stories'] = df_housing['stories'].astype(np.int8)
df_housing['parking'] = df_housing['parking'].astype(np.int8)
df_housing['mainroad'] = df_housing['mainroad'].astype("category")
df_housing['guestroom'] = df_housing['guestroom'].astype("category")
df_housing['basement'] = df_housing['basement'].astype("category")
df_housing['hotwaterheating'] = df_housing['hotwaterheating'].astype("category")
df_housing['airconditioning'] = df_housing['airconditioning'].astype("category")
df_housing['prefarea'] = df_housing['prefarea'].astype("category")
df_housing['furnishingstatus'] = df_housing['furnishingstatus'].astype("category")

In [42]:
df_housing.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    int32   
 1   area              545 non-null    int16   
 2   bedrooms          545 non-null    int8    
 3   bathrooms         545 non-null    int8    
 4   stories           545 non-null    int8    
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    int8    
 11  prefarea          545 non-null    category
 12  furnishingstatus  545 non-null    category
dtypes: category(7), int16(1), int32(1), int8(4)
memory usage: 10.0 KB
